### DISCUSSION TITLE
MacPorts build of c3 on intel macOS Catalina osx 10.15.7

### DISCUSSION THREAD

**Discussion by c3241220** [c3lang/c3c#1701] [General]
Hello.  I came across C3 and thought I would give it a try.
I found there's no prebuilt binary for my platform.
Unfortunately brew is not supported on my older version of OSX so I use https://www.macports.org

I followed a variation of the instructions https://c3-lang.org/getting-started/compile
I had previously installed llvm-17 via macports so I thought I would be ok, but I had a few issues to work through.
Leaving this here for anyone else who may find the details useful and perhaps save a little time.

## TL;DR

Although I already had llvm-17 on my system, I was not able to build c3c from a clone of the repository **until** doing the following steps

* export LLVM_DIR=/opt/local/libexec/llvm-17/lib/cmake/llvm
* sudo port install clang-17

Details on what I did follow

## First Attempt

After cloning the repository and creating a **build/** directory the command "cmake .." failed with an error
```
CMake Error at CMakeLists.txt:149 (find_package):
  Could not find a package configuration file provided by "LLVM" with any of
  the following names:

    LLVMConfig.cmake
    llvm-config.cmake

  Add the installation prefix of "LLVM" to CMAKE_PREFIX_PATH or set
  "LLVM_DIR" to a directory containing one of the above files.  If "LLVM"
  provides a separate development package or SDK, be sure it has been
  installed.
```

Based on the output of "port info llvm-17" and "port contents llvm-17" I determined
I indeed had llvm-17 already, but didn't know the proper setting of **LLVM_DIR**.

After some searching I came across a mention of LLVM_DIR at https://forums.gentoo.org/viewtopic-t-1168721.html
which helped me find **/opt/local/libexec/llvm-17/lib/cmake/llvm** on my system.

## Second Attempt

After removing and recreating the **build/** directory and explicitly setting
```
export LLVM_DIR=/opt/local/libexec/llvm-17/lib/cmake/llvm
```

the "cmake .." command failed with different errors

```
-- Configuring done (5.8s)
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
LLD_COFF
    linked by target "c3c_wrappers" in directory /Volumes/Experiments/c3c
    linked by target "c3c" in directory /Volumes/Experiments/c3c
LLD_COMMON
    linked by target "c3c_wrappers" in directory /Volumes/Experiments/c3c
    linked by target "c3c" in directory /Volumes/Experiments/c3c
LLD_ELF
    linked by target "c3c_wrappers" in directory /Volumes/Experiments/c3c
    linked by target "c3c" in directory /Volumes/Experiments/c3c
LLD_MACHO
    linked by target "c3c_wrappers" in directory /Volumes/Experiments/c3c
    linked by target "c3c" in directory /Volumes/Experiments/c3c
LLD_MINGW
    linked by target "c3c_wrappers" in directory /Volumes/Experiments/c3c
    linked by target "c3c" in directory /Volumes/Experiments/c3c
LLD_WASM
    linked by target "c3c_wrappers" in directory /Volumes/Experiments/c3c
    linked by target "c3c" in directory /Volumes/Experiments/c3c
```

and as expected the "cmake --build ." step quickly failed

```
% cmake --build .
[  1%] Building C object CMakeFiles/miniz.dir/dependencies/miniz/miniz.c.o
[  2%] Linking C static library libminiz.a
[  2%] Built target miniz
[  3%] Building CXX object CMakeFiles/c3c_wrappers.dir/wrapper/src/wrapper.cpp.o
/Volumes/Experiments/c3c/wrapper/src/wrapper.cpp:11:10: fatal error: 'lld/Common/CommonLinkerContext.h' file not found
#include "lld/Common/CommonLinkerContext.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

More searching turned up https://github.com/c3lang/c3c/issues/399
which confirmed my belief that I was likely missing **lld** or it was in an unexpected place.
On a whim I tried searching for the header file in the error message.

```
% find /opt -name CommonLinkerContext.h
/opt/local/libexec/llvm-15/include/lld/Common/CommonLinkerContext.h
/opt/local/libexec/llvm-18/include/lld/Common/CommonLinkerContext.h
```

Very odd... why did I have the file for llvm-15 and llvm-18 but not llvm-17?
And why did I have those at all?   Well some weeks ago I installed various
versions of clang in an effort to build zig from source, which made me believe
I probably needed **clang-17** in addtion to **llvm-17**

Running "sudo port install clang-17" provided me with
/opt/local/libexec/llvm-17/include/lld/Common/CommonLinkerContext.h
as expected.

## Third Attempt

This time I was able to build c3c successfully

```
% export LLVM_DIR=/opt/local/libexec/llvm-17/lib/cmake/llvm
% mkdir build
% cd build
% cmake ..
...
-- Configuring done (10.2s)
-- Generating done (1.5s)
-- Build files have been written to: /Volumes/Experiments/c3c/build

% cmake --build .
[  1%] Building C object CMakeFiles/miniz.dir/dependencies/miniz/miniz.c.o
[  2%] Linking C static library libminiz.a
[  2%] Built target miniz
...
[100%] Linking CXX executable c3c
Copying sanitizer runtime libraries to output directory
[100%] Built target c3c
```

This c3c build and ran the "hello_world.c3" as described at https://c3-lang.org/getting-started/hello-world

## otool output

In response to @lerno, here is the output of "otool -L hello_world" 

```
% otool -L hello_world
hello_world:
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
```

Also a "hello_world.dSYM" was built as well with the following contents

```
% tree hello_world.dSYM/
hello_world.dSYM/
└── Contents
    ├── Info.plist
    └── Resources
        └── DWARF
            └── hello_world

3 directories, 2 files

```

**Comment by lerno** [comment]
It would be very useful if you would like to contribute this to the build / install instructions. I've had similar "fun" times when using various versions of LLVM on MacOS. If you use macports, can you see what dynamic library dependencies the resulting c3c binary has?

**Reply by c3241220** [reply]
I will create a PR with some recommended steps for macports to the bottom of https://c3-lang.org/getting-started/compile
I've added the "otool -L" output for the binary along with a "tree" of the hello_world.dSYM that was also generated.

**Reply by lerno** [reply]
Another contributor was interesting in creating a separate INSTALL.md for the git repo, so we can add it to that one too at a later point.

**Reply by c3241220** [reply]
Created https://github.com/c3lang/c3-web/pull/92

**Comment by barracuda156** [comment]
@c3241220 I am also interested in the port, but with a variant not to use LLVM at all.
